src/odb.c

```diff
@@ -811,11 +850,38 @@ int git_odb_read_header(size_t *len_p, git_otype *type_p, git_odb *db, const git
 	return error;
 }
 
+static int odb_read_header_1(
+	size_t *len_p, git_otype *type_p, git_odb *db,
+	const git_oid *id, bool only_refreshed)
+{
+	size_t i;
+	int error = GIT_PASSTHROUGH;
+	git_otype ht;
+
+	if (!only_refreshed && (ht = odb_hardcoded_type(id)) != GIT_OBJ_BAD) {
+		*type_p = ht;
+		*len_p = 0;
+		return 0;
+	}
+
+	for (i = 0; i < db->backends.length && error < 0; ++i) {
+		backend_internal *internal = git_vector_get(&db->backends, i);
+		git_odb_backend *b = internal->backend;
+
+		if (only_refreshed && !b->refresh)
+			continue;
+
+		if (b->read_header != NULL)
+			error = b->read_header(len_p, type_p, b, id);
```

Comment:
We will now stop this loop immediately and return `E_PASSTHROUGH` if any of the `read_header`s returned `E_PASSTHROUGH`.  I feel like the correct thing to do here would be to remember that one of the backends returned `E_PASSTHROUGH`, but continue.  One of the other backends may be the one that actually has this object.

This lets that later-ordered backend service the `read_header` request, on the assumption that `read_header` is cheaper than the full `read`.